src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -209,13 +255,21 @@ rawPageValueAt page@(RawPage off ba) i = do
     start = fromIntegral (P.unsafeIndex offs i) :: Int
     end   = fromIntegral (P.unsafeIndex offs (i + 1)) :: Int
 
-rawPageSingleValue :: RawPage -> (P.Vector Word8, Word32)
-rawPageSingleValue page@(RawPage off ba) =
-    if end > 4096
-    then (P.Vector (mul2 off + fromIntegral start) (4096 - fromIntegral start) ba, end - 4096)
-    else (P.Vector (mul2 off + fromIntegral start) (fromIntegral end - fromIntegral start) ba, 0)
+rawPageSingleValuePrefix :: RawPage -> P.Vector Word8
+rawPageSingleValuePrefix page@(RawPage off ba) =
+    P.Vector (mul2 off + fromIntegral start)
+             (fromIntegral prefix_end - fromIntegral start)
+             ba
   where
     (start, end) = rawPageValueOffsets1 page
+    prefix_end   = min 4096 end
```

Comment:
Is the end offset the offset of the last byte, or the offset after the last byte? I think it's the latter, but just checking. Maybe we should specify this in the format doc

Reply:
Good Q. I'll check and yes we should make sure the docs are clear on this.

Reply:
It has to be past, otherwise there wouldn't be a a way to represent empty keys/values.

Reply:
Yes, all the offsets represent [open, closed) intervals. That also allows the representation where we only store the offset of the first byte in each string, since the subsequent entry provides the exclusive upper bound.

Reply:
The existing text says:

> Entry i (indexed from 0) gives the offset within the page of the start of the key byte string. Entry i+1 therefore gives the offset one past each key.

But I'm adding another sentence just to be redundant and explicit.